cedar-lean/DiffTest/Parser.lean

```diff
@@ -345,29 +345,41 @@ def jsonToActionScope (json : Lean.Json) : ParseResult ActionScope := do
     .ok (.actionScope (.eq euid))
   | tag => .error s!"jsonToActionScope: unknown tag {tag}"
 
-def jsonToPolicy (json : Lean.Json) : ParseResult Policy := do
-  let id ← getJsonField json "id" >>= jsonToString
+def jsonToTemplate (json : Lean.Json) : ParseResult Template := do
   let effect ← getJsonField json "effect" >>= jsonToEffect
-  let principalConstraint ← getJsonField json "principal_constraint" >>= (getJsonField · "constraint") >>= jsonToScope
+  let principalConstraint ← getJsonField json "principal_constraint" >>= (getJsonField · "constraint") >>= jsonToScopeTemplate "?principal"
   let actionConstraint ← getJsonField json "action_constraint" >>= jsonToActionScope
-  let resourceConstraint ← getJsonField json "resource_constraint" >>= (getJsonField · "constraint") >>= jsonToScope
+  let resourceConstraint ← getJsonField json "resource_constraint" >>= (getJsonField · "constraint") >>= jsonToScopeTemplate "?resource"
   let condition ← getJsonField json "non_head_constraints" >>= jsonToExpr
   .ok {
-    id := id
```

Comment:
do templates not have ids

Reply:
Templates do have ids, but I decided that it was more straightforward to store them as map keys (see defn below) rather than in the data structure directly.
```
abbrev Templates := Map TemplateID Template
```
The current parsing code pulls out the ids in the `jsonToPolicies` function.

Reply:
In case it matters, we decided to keep IDs as part of the `Policy` datatype: https://github.com/cedar-policy/cedar-spec/blob/d17739d370bccc9836f59053a56b558fc9803a70/cedar-lean/Cedar/Spec/Policy.lean#L57

This is largely because it made the `isAuthorized` code and proofs nicer, in particular the proof that authorization decisions are independent of duplicates and evaluation order (https://github.com/cedar-policy/cedar-spec/blob/d17739d370bccc9836f59053a56b558fc9803a70/cedar-lean/Cedar/Thm/Authorization.lean#L90).

Not sure if we plan to do any proofs re templates, but if we do, I suspect that having a list of templates with embedded IDs may be easier to work with than maps.

Reply:
Makes sense! I'll leave it as-is for now, and we can reconsider if/when we decide to do proofs with templates.